xformers/components/attention/base.py

```diff
@@ -53,6 +53,10 @@ def __init__(self, dropout: Optional[float] = None, *args, **kwargs):
         # so that the MHA wrapper should skip it
         self.requires_skip_multi_head = False
 
+        # Whether this attention mechanism supports attention masks
+        self.supports_attention_mask = True
+        self.supports_key_padding_mask = False
```

Comment:
Ideally, I think I would prefer if we only support `attention_mask`, and then support the "optimized" case of `key_padding_mask` internally by checking the strides of the `attention_mask` Tensor.
The `key_padding_mask` case is actually just assuming that the stride of dimension -2 of the `attention_mask` is 0. So the converting `key_padding_mask` to a `attention_mask` can be done efficiently via
```python
attn_mask = key_padding_mask[:, None, None, :].expand(batch, heads, query_len, key_len)
```

Indeed, are we sure that all our implementations support both `attention_mask` *and* `key_padding_mask` at the same time?

Reply:
Agree on the simpler interface, it's a bit confusing right now. One trick is that for some attentions (Nystrom for instance) there's no attention mask but there's a possible key padding mask (which can then generate a mask of sorts indeed, but in that case the dimensions will differ). I think we could do a second pass on that indeed, thoughts @dianaml0  ?

Reply:
I agree ideally we could have this simpler approach but I ended up adding a key padding mask argument separately for Nystrom since Nystrom can't apply attention masks. I couldn't come up with a way around that but maybe there's a better way

Reply:
hmm so ideally we would just support key padding for nystrom and not make it visible everywhere ? right now there's a global flag because this happens in the MHA, but maybe that we could more this one level down and not expose this for other mechanisms ?

Reply:
else it feels like an abstraction leak on our end, not perfect.. could be worth some more thinking